### PR TITLE
Removed OSCache. Using EHCache and Guava instead.

### PR DESCRIPTION
**Pull request by theoretical-olive**

### CODE REVIEW

File: gateway/pom.xml [hunk 1]
```diff
@@ -62,11 +62,11 @@
             <artifactId>dropwizard-assets</artifactId>
             <version>${dropwizard.version}</version>
         </dependency>
-        <dependency>
-            <groupId>org.ehcache</groupId>
-            <artifactId>ehcache</artifactId>
-            <version>${ehcache.version}</version>
-        </dependency>
+        <!--<dependency>-->
```
Comment (puneetjaiswal) [inline gateway/pom.xml]: let's remove this.

File: gateway-ha/src/test/java/com/lyft/data/gateway/ha/router/TestQueryHistoryManager.java [hunk 2]
```diff
@@ -42,8 +41,8 @@ public void testSubmitAndFetchQueryHistory() {
     queryDetail.setUser("test@ea.com");
     queryDetail.setQueryText("select 1");
     for (int i = 0; i < 2; i++) {
-      queryDetail.setQueryId(String.valueOf(Time.absoluteTime()));
-      queryDetail.setCaptureTime(Time.absoluteTime());
+      queryDetail.setQueryId(String.valueOf(new Timestamp(System.currentTimeMillis()).getTime()));
```
Comment (puneetjaiswal) [inline gateway-ha/src/test/java/com/lyft/data/gateway/ha/router/TestQueryHistoryManager.java]: `timestamp.getTime()` returns same value as `System.currentTimeMillis()`

Comment (puneetjaiswal) [reply gateway-ha/src/test/java/com/lyft/data/gateway/ha/router/TestQueryHistoryManager.java]: `String.valueOf(new Timestamp(System.currentTimeMillis()).getTime())` can be replaced as `String.valueOf(System.currentTimeMillis())`

File: gateway/src/main/java/com/lyft/data/gateway/router/RoutingManager.java [hunk 3]
```diff
@@ -1,62 +1,52 @@
...
-                        .heap(1000, EntryUnit.ENTRIES)
-                        .offheap(100, MemoryUnit.MB)
-                        .disk(1, MemoryUnit.GB, true)))
-            .build(true);
-    this.queryIdBackendCache =
-        persistentCacheManager.getCache(
-            "queryIdBackendPersistentCache", String.class, String.class);
+    queryIdBackendCache =
+        CacheBuilder.newBuilder()
+            .maximumSize(1000)
```
Comment (puneetjaiswal) [inline gateway/src/main/java/com/lyft/data/gateway/router/RoutingManager.java]: let's increase this to 5K?

File: gateway/src/main/java/com/lyft/data/gateway/router/RoutingManager.java [hunk 4]
```diff
@@ -1,62 +1,52 @@
...
+            .expireAfterWrite(30, TimeUnit.MINUTES)
```
Comment (puneetjaiswal) [inline gateway/src/main/java/com/lyft/data/gateway/router/RoutingManager.java]: let's replace this with `expireAfterAccess`?

**Comment by puneetjaiswal** [comment]
Few minor comments. Looks good otherwise. Please create a PR to `lyft/presto-gateway` after updating.

File: gateway/src/main/java/com/lyft/data/gateway/router/RoutingManager.java [hunk 5]
```diff
@@ -1,62 +1,52 @@
...
-                        .heap(1000, EntryUnit.ENTRIES)
-                        .offheap(100, MemoryUnit.MB)
-                        .disk(1, MemoryUnit.GB, true)))
-            .build(true);
-    this.queryIdBackendCache =
-        persistentCacheManager.getCache(
-            "queryIdBackendPersistentCache", String.class, String.class);
+    queryIdBackendCache =
+        CacheBuilder.newBuilder()
+            .maximumSize(5000)
```
Comment (puneetjaiswal) [inline gateway/src/main/java/com/lyft/data/gateway/router/RoutingManager.java]: 5k seems lesser given our workload. Let's set this to 10K.

File: gateway/src/main/java/com/lyft/data/gateway/router/RoutingManager.java [hunk 6]
```diff
@@ -68,7 +58,9 @@ public void setBackendForQueryId(String queryId, String backend) {
   }
 
   /**
-   * Performs routing to an adhoc backend.
+   * Performs routing to an adhoc backen
+   *
+   * <p>d.
```
Comment (puneetjaiswal) [inline gateway/src/main/java/com/lyft/data/gateway/router/RoutingManager.java:63]: typo here.

File: gateway/src/main/java/com/lyft/data/gateway/router/RoutingManager.java [hunk 7]
```diff
@@ -105,10 +97,14 @@ public String provideBackendForRoutingGroup(String routingGroup) {
...
+    String backendAddress = null;
+    try {
+      backendAddress = queryIdBackendCache.get(queryId);
+    } catch (ExecutionException e) {
+      log.error("Exception while loading queryId from cache {}", e.getLocalizedMessage());
+    }
     if (Strings.isNullOrEmpty(backendAddress)) {
       log.error("Could not find mapping for query id {}", queryId);
-      // for now fall back to the first backend as default
       backendAddress = findBackendForUnknownQueryId(queryId);
```
Comment (puneetjaiswal) [inline gateway/src/main/java/com/lyft/data/gateway/router/RoutingManager.java]: this case is already handled in line 47. We can get rid of this line.

File: gateway-ha/README.md [hunk 8]
```diff
@@ -27,3 +65,233 @@ Step 3: Add below program argument to class `HaGatewayLauncher` and debug in IDE
...
+## Gateway HA API
+
+### Get all backends behind the gateway
+
+`curl -X GET prestogateway.lyft.com/gateway/backend/all | python -m json.tool`
+```
+[
+    {
+        "active": true,
+        "includeInRouter": true,
```
Comment (puneetjaiswal) [inline gateway-ha/README.md]: let's remove `includeInRouter` and `localPort` options from here.

**Review by puneetjaiswal — COMMENTED** [review]
Very nice!

File: gateway-ha/README.md (hunk 8 above)
Comment (theoretical-olive) [reply gateway-ha/README.md]: From the post-call? The response body has those two options.

**Pull request closed without merging**